lib/hal/portable/st/aws_hal_perf_cntr.c

```diff
@@ -0,0 +1,237 @@
+*
+* Permission is hereby granted, free of charge, to any person obtaining a copy
+* of this software and associated documentation files (the "Software"), to deal
+* in the Software without restriction, including without limitation the rights
+* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+* copies of the Software, and to permit persons to whom the Software is
+* furnished to do so, subject to the following conditions:
+*
+* The above copyright notice and this permission notice shall be included in
+* all copies or substantial portions of the Software.
+*
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+* SOFTWARE.
+*
+* http://aws.amazon.com/freertos
+* http://www.FreeRTOS.org
+*/
+
+/**
+* @file aws_hal_perf_cntr.c
+* @brief File of the implementation for performance counter APIs calling STM drivers.
+*/
+
+#include <stdint.h>
+
+/* ST HAL APIs to be called in this file. */
+#include "stm32l4xx_hal.h"
+
+/* API definition to be implemented in this file. */
+#include "aws_hal_perf_cntr.h"
+
+/* Load FreeRTOS user defined performance counter configurations. */
+#include "FreeRTOSConfig.h"
+
+/**
+ * @brief Default performance counter frequency.
+ *
+ * Default is set to 10MHz ==> (uint32_t) 0xffffffff / 10MHz = 429.4967s.
+ * So roughly every 7 mins, kernel needs to serve an overflow IRQ.
+ */
+#define PERF_COUNTER_FREQ_DEFAULT           ( 10000000 )
+
+/**
+ * @brief Maximum performance counter frequency.
+ *
+ * Maximum frequency for APB1 is 80MHz. Kernel needs to serve and overflow
+ * IRQ in each ~53.6870s.
+ */
+#define PERF_COUNTER_FREQ_MAX               ( 80000000 )
+
+
+/**
+ * User suggested performance counter frequency.
+ *
+ * @warning The value of configHAL_PERF_COUNTER_FREQ needs to be smaller than
+ * TIM clock.
+ *
+ * @warning TIM prescaler only takes integer. configHAL_PERF_COUNTER_FREQ can
+ * be any value, but prescaler is (uint32_t)(PCLK1 frequency / configHAL_PERF_COUNTER_FREQ).
+ */
+#ifdef configHAL_PERF_COUNTER_FREQ
+    #define HAL_PERF_COUNTER_FREQ       ( configHAL_PERF_COUNTER_FREQ )
+#else
+    #define HAL_PERF_COUNTER_FREQ       ( PERF_COUNTER_FREQ_DEFAULT )
+#endif
+
+/**
+ * @brief Timer width, counter period, and loading value.
+ */
+#define HW_TIMER_32_WIDTH           ( sizeof( uint32_t ) * 8 )
+#define HW_TIMER_32_CONST_PERIOD    ( UINT32_MAX )
+#define HW_TIMER_32_LOADING_VALUE   ( 0x0UL )
+
+
+/*-----------------------------------------------------------*/
+
+static TIM_HandleTypeDef xTimerHandle = { 0 };
+static uint64_t ullTimerOverflow = 0;
```

Comment:
This should just be 32 bit, and then you cast it before shifting it in get_value.